File: examples/python/edit_showcase_words.py
from __future__ import annotations

import json
import shutil
import subprocess
import tempfile
from pathlib import Path

from officemd import (
    DocxPatch,
    DocxTextScope,
    MatchPolicy,
    PptxPatch,
    PptxTextScope,
    ScopedDocxReplace,
    ScopedPptxReplace,
    ScopedXlsxReplace,
    TextReplace,
    XlsxPatch,
    XlsxSheetRename,
    XlsxTextScope,
    extract_ir_json,
    extract_sheet_names,
    markdown_from_bytes,
    patch_docx,
    patch_pptx,
    patch_xlsx,
    patch_xlsx_with_report,
)

REPO_ROOT = Path(__file__).resolve().parents[2]
DATA_DIR = REPO_ROOT / "examples" / "data"
OUT_DIR = REPO_ROOT / "examples" / "out"


def _check_with_libreoffice(path: Path) -> tuple[bool, str]:
    soffice = shutil.which("soffice") or shutil.which("libreoffice")
    if not soffice:
        return False, "LibreOffice CLI not found"

    with tempfile.TemporaryDirectory(prefix="officemd-lo-profile-") as profile_dir, tempfile.TemporaryDirectory(
        prefix="officemd-lo-out-"
    ) as out_dir:
        cmd = [
            soffice,
            f"-env:UserInstallation=file://{Path(profile_dir).resolve()}",
            "--headless",
            "--convert-to",
            "pdf",
            "--outdir",
            out_dir,
            str(path),
        ]
        completed = subprocess.run(cmd, capture_output=True, text=True, check=False)
        produced = sorted(Path(out_dir).glob("*.pdf"))
        output = (completed.stdout + completed.stderr).strip()
        return completed.returncode == 0 and bool(produced), output or "LibreOffice conversion failed"


def _docx_patch() -> DocxPatch:
    return DocxPatch(
        set_core_title="Edited DOCX Showcase From Python",
        scoped_replacements=[
            ScopedDocxReplace(
                DocxTextScope.HEADERS,
                TextReplace("OOXML Showcase Header", "OfficeMD Showcase Header — edited from Python"),
            ),
            ScopedDocxReplace(
                DocxTextScope.BODY,
                TextReplace("Quarterly Operations Summary", "Quarterly Operations Summary — edited from Python"),
            ),
            ScopedDocxReplace(
                DocxTextScope.COMMENTS,
                TextReplace(
                    "Example DOCX comment captured as markdown footnote.",
                    "Edited DOCX comment from Python patch API.",
                ),
            ),
            ScopedDocxReplace(
                DocxTextScope.METADATA_APP,
                TextReplace("OfficeMD", "OfficeMD Python Example"),
            ),
            ScopedDocxReplace(
                DocxTextScope.METADATA_CUSTOM,
                TextReplace("showcase", "showcase-python"),
            ),
        ],
    )


def _pptx_patch() -> PptxPatch:
    return PptxPatch(
        set_core_title="Edited PPTX Showcase From Python",
        scoped_replacements=[
            ScopedPptxReplace(
                PptxTextScope.ALL_TEXT,
                TextReplace(
                    "Quarterly Review",
                    "Quarterly Review — edited from Python",
                    match_policy=MatchPolicy.EXACT,
                ),
            ),
            ScopedPptxReplace(
                PptxTextScope.COMMENTS,
                TextReplace(
                    "Add one slide on operating margin.",
                    "Edited PPTX comment from Python patch API.",
                ),
            ),
            ScopedPptxReplace(
                PptxTextScope.COMMENT_AUTHORS,
                TextReplace("Alice", "Python Reviewer"),
            ),
            ScopedPptxReplace(
                PptxTextScope.METADATA_APP,
                TextReplace("OfficeMD", "OfficeMD Python Example"),
            ),
        ],
    )


def _xlsx_patch() -> XlsxPatch:
    return XlsxPatch(
        set_core_title="Edited XLSX Showcase From Python",
        rename_sheets=[XlsxSheetRename("Sales", "Sales Python")],
        scoped_replacements=[
            ScopedXlsxReplace(
                XlsxTextScope.ALL_TEXT,
                TextReplace("North", "North Python"),
            ),
            ScopedXlsxReplace(
                XlsxTextScope.COMMENTS,
                TextReplace("Review", "Reviewed from Python"),
            ),
            ScopedXlsxReplace(
                XlsxTextScope.METADATA_APP,
                TextReplace("OfficeMD", "OfficeMD Python Example"),
            ),
            ScopedXlsxReplace(
                XlsxTextScope.METADATA_CUSTOM,
                TextReplace("showcase", "showcase-python"),
            ),
        ],
    )


def _summarize(path: Path, fmt: str) -> dict[str, object]:
    content = path.read_bytes()
    markdown = markdown_from_bytes(content, format=fmt, include_document_properties=True)
    ir = json.loads(extract_ir_json(content, format=fmt))
    libreoffice_ok, libreoffice_output = _check_with_libreoffice(path)
    result: dict[str, object] = {
        "path": str(path.relative_to(REPO_ROOT)),
        "markdown_excerpt": "\n".join(markdown.splitlines()[:18]),
        "libreoffice_ok": libreoffice_ok,
        "libreoffice_output": libreoffice_output,
    }
    if fmt == "docx":
        result["core_title"] = ir["properties"]["core"].get("title")
        result["body_title"] = ir["sections"][0]["blocks"][0]["Paragraph"]["inlines"][0]["Text"]
        result["header_text"] = ir["sections"][1]["blocks"][0]["Paragraph"]["inlines"][0]["Text"]
        result["has_comment"] = "Edited DOCX comment from Python patch API." in markdown
    elif fmt == "pptx":
        result["core_title"] = (ir.get("properties") or {}).get("core", {}).get("title")
        result["slide_1_title"] = ir["slides"][0]["title"]
        result["has_comment"] = "Edited PPTX comment from Python patch API." in markdown
    else:
        result["core_title"] = (ir.get("properties") or {}).get("core", {}).get("title")
        result["sheet_names"] = extract_sheet_names(content)
        result["has_python_edit"] = "North Python" in markdown or "Reviewed from Python" in markdown
    return result


def main() -> None:
    OUT_DIR.mkdir(parents=True, exist_ok=True)

    print("Plan:")
    print("1. Load examples/data/showcase.docx, showcase.pptx, and showcase.xlsx")
    print("2. Patch OOXML parts directly via officemd.patch_docx / patch_pptx / patch_xlsx")
    print("3. Showcase metadata/comment-author scopes plus ALL_TEXT semantics")
    print("4. Verify edited files via officemd markdown/IR extraction")
    print("5. Verify LibreOffice can still open them via headless PDF conversion")
    print()

    docx_out = OUT_DIR / "showcase_edited.docx"
    docx_out.write_bytes(patch_docx((DATA_DIR / "showcase.docx").read_bytes(), _docx_patch()))

    pptx_out = OUT_DIR / "showcase_edited.pptx"
    pptx_out.write_bytes(patch_pptx((DATA_DIR / "showcase.pptx").read_bytes(), _pptx_patch()))

    xlsx_out = OUT_DIR / "showcase_edited.xlsx"
    xlsx_result = patch_xlsx_with_report((DATA_DIR / "showcase.xlsx").read_bytes(), _xlsx_patch())
    xlsx_out.write_bytes(xlsx_result.content)

    print("DOCX result:")
    print(json.dumps(_summarize(docx_out, "docx"), indent=2, ensure_ascii=False))
    print()
    print("PPTX result:")
    print(json.dumps(_summarize(pptx_out, "pptx"), indent=2, ensure_ascii=False))
    print()
    print("XLSX result:")
    xlsx_summary = _summarize(xlsx_out, "xlsx")
    xlsx_summary["report"] = {
        "parts_scanned": xlsx_result.report.parts_scanned,
        "parts_modified": xlsx_result.report.parts_modified,
        "replacements_applied": xlsx_result.report.replacements_applied,
    }
    print(json.dumps(xlsx_summary, indent=2, ensure_ascii=False))


if __name__ == "__main__":
    main()

File: examples/rust/src/bin/edit_showcase_words.rs
use officemd_core::{
    DocxPatch, DocxTextScope, MatchPolicy, PptxPatch, PptxTextScope, ScopedDocxReplace,
    ScopedPptxReplace, TextReplace, patch_docx, patch_pptx,
};
use serde_json::json;
use std::fs;
use std::path::{Path, PathBuf};
use std::process::Command;

fn repo_root() -> PathBuf {
    PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .parent()
        .expect("examples dir")
        .to_path_buf()
}

fn check_libreoffice(path: &Path) -> (bool, String) {
    let soffice = ["soffice", "libreoffice"]
        .into_iter()
        .find_map(|name| which::which(name).ok());
    let Some(soffice) = soffice else {
        return (false, "LibreOffice CLI not found".to_string());
    };

    let profile_dir = tempfile::tempdir().expect("profile tempdir");
    let out_dir = tempfile::tempdir().expect("out tempdir");
    let output = Command::new(soffice)
        .arg(format!(
            "-env:UserInstallation=file://{}",
            profile_dir.path().display()
        ))
        .arg("--headless")
        .arg("--convert-to")
        .arg("pdf")
        .arg("--outdir")
        .arg(out_dir.path())
        .arg(path)
        .output()
        .expect("run soffice");
    let produced = fs::read_dir(out_dir.path())
        .expect("read out dir")
        .flatten()
        .any(|entry| entry.path().extension().and_then(|s| s.to_str()) == Some("pdf"));
    let text = format!(
        "{}{}",
        String::from_utf8_lossy(&output.stdout),
        String::from_utf8_lossy(&output.stderr)
    )
    .trim()
    .to_string();
    (output.status.success() && produced, text)
}

fn main() {
    let root = repo_root();
    let data_dir = root.join("data");
    let out_dir = root.join("out");
    fs::create_dir_all(&out_dir).expect("create out dir");

    println!("Plan:");
    println!("1. Load examples/data/showcase.docx and showcase.pptx");
    println!("2. Patch OOXML parts directly with officemd_core::patch_docx / patch_pptx");
    println!("3. Verify the edited files with existing officemd extractors");
    println!("4. Open-check via LibreOffice headless PDF conversion");
    println!();

    let docx_bytes = fs::read(data_dir.join("showcase.docx")).expect("read docx");
    let edited_docx = patch_docx(
        &docx_bytes,
        &DocxPatch {
            set_core_title: Some("Edited DOCX Showcase From Rust".to_string()),
            replace_body_title: None,
            scoped_replacements: vec![
                ScopedDocxReplace {
                    scope: DocxTextScope::Headers,
                    replace: TextReplace::all(
                        "OOXML Showcase Header",
                        "OfficeMD Showcase Header — edited from Rust",
                    ),
                },
                ScopedDocxReplace {
                    scope: DocxTextScope::Body,
                    replace: TextReplace::first(
                        "Quarterly Operations Summary",
                        "Quarterly Operations Summary — edited from Rust",
                    ),
                },
                ScopedDocxReplace {
                    scope: DocxTextScope::Comments,
                    replace: TextReplace::all(
                        "Example DOCX comment captured as markdown footnote.",
                        "Edited DOCX comment from Rust patch API.",
                    )
                    .with_match_policy(MatchPolicy::Exact),
                },
            ],
        },
    )
    .expect("patch docx");
    let docx_out = out_dir.join("showcase_edited_rust.docx");
    fs::write(&docx_out, &edited_docx).expect("write docx");
    let docx_md = officemd_docx::markdown_from_bytes(&edited_docx).expect("markdown docx");
    let docx_ir: serde_json::Value =
        serde_json::from_str(&officemd_docx::extract_ir_json(&edited_docx).expect("ir docx"))
            .expect("parse ir docx");
    let (docx_ok, docx_lo) = check_libreoffice(&docx_out);

    let pptx_bytes = fs::read(data_dir.join("showcase.pptx")).expect("read pptx");
    let edited_pptx = patch_pptx(
        &pptx_bytes,
        &PptxPatch {
            set_core_title: Some("Edited PPTX Showcase From Rust".to_string()),
            scoped_replacements: vec![
                ScopedPptxReplace {
                    scope: PptxTextScope::AllText,
                    replace: TextReplace::first(
                        "Quarterly Review",
                        "Quarterly Review — edited from Rust",
                    ),
                },
                ScopedPptxReplace {
                    scope: PptxTextScope::Comments,
                    replace: TextReplace::all(
                        "Add one slide on operating margin.",
                        "Edited PPTX comment from Rust patch API.",
                    ),
                },
            ],
        },
    )
    .expect("patch pptx");
    let pptx_out = out_dir.join("showcase_edited_rust.pptx");
    fs::write(&pptx_out, &edited_pptx).expect("write pptx");
    let pptx_md = officemd_pptx::markdown_from_bytes(&edited_pptx).expect("markdown pptx");
    let pptx_ir: serde_json::Value =
        serde_json::from_str(&officemd_pptx::extract_ir_json(&edited_pptx).expect("ir pptx"))
            .expect("parse ir pptx");
    let (pptx_ok, pptx_lo) = check_libreoffice(&pptx_out);

    println!("DOCX result: {}", docx_out.display());
    println!(
        "{}",
        serde_json::to_string_pretty(&json!({
            "core_title": docx_ir["properties"]["core"]["title"],
            "body_title": docx_ir["sections"][0]["blocks"][0]["Paragraph"]["inlines"][0]["Text"],
            "header_text": docx_ir["sections"][1]["blocks"][0]["Paragraph"]["inlines"][0]["Text"],
            "markdown_has_comment": docx_md.contains("Edited DOCX comment from Rust patch API."),
            "libreoffice_ok": docx_ok,
            "libreoffice_output": docx_lo,
        }))
        .unwrap()
    );
    println!();

    println!("PPTX result: {}", pptx_out.display());
    println!(
        "{}",
        serde_json::to_string_pretty(&json!({
            "core_title": pptx_ir["properties"]["core"]["title"],
            "slide_1_title": pptx_ir["slides"][0]["title"],
            "markdown_has_comment": pptx_md.contains("Edited PPTX comment from Rust patch API."),
            "libreoffice_ok": pptx_ok,
            "libreoffice_output": pptx_lo,
        }))
        .unwrap()
    );
}
